webapp/src/js/enketo/medic-xpath-extensions.js

```diff
@@ -56,10 +61,47 @@ const parseTimestampToDate = (value) => {
   const timestamp = parseInt(getValue(value));
 
   if (isNaN(timestamp)) {
-    return { t: 'str', v: '' };
+    return XPR.string('');
   }
 
-  return { t:'date', v: new Date(timestamp) };
+  return XPR.date(new Date(timestamp));
+};
+
+const asString = (r) => {
+  return r.t === 'arr' ?
+    r.v.length && !(r.v[0] === null || r.v[0] === undefined) ? r.v[0].textContent || '' : '' :
+    r.v.toString();
+};
+
+// Based on https://github.com/enketo/openrosa-xpath-evaluator/blob/3bfcb493ec01cf84f55e254a096a31e5be01de15/src/openrosa-extensions.js#L547
+const asDate = (r) => {
+  const dateSinceUnixEpoch = (days) => {
+    // Create a date at 00:00:00 1st Jan 1970 _in the current timezone_
+    const date = new Date(1970, 0, 1);
+    date.setDate(1 + days);
+    return date;
+  };
+  switch(r.t) {
+  case 'bool':
+    return new Date(NaN);
+  case 'date':
+    return r.v;
+  case 'num':
+    return dateSinceUnixEpoch(r.v);
+  case 'arr':
+  default: {
+    r = asString(r);
+    if(RAW_NUMBER.test(r)) {
+      return dateSinceUnixEpoch(parseInt(r, 10));
+    }
+    const rMoment = moment(r);
+    if(rMoment.isValid()) {
+      const time = `${rMoment.format('YYYY-MM-DD')}T00:00:00.000${getTimezoneOffsetAsTime(new Date(r))}`;
```

Comment:
Okay so this seems to trip up on the weird logic from openrosa-xpath-evaluator where it blanks out the time (but only for date strings of a certain format...).  If I just return the moment object here with no manipulations, it changes the result for input like `2014-10-01T11:11:11.111`:
```
Expected: Wed Oct 01 2014 00:00:00 GMT-0500 (Central Daylight Time)
Actual : Wed Oct 01 2014 11:11:11 GMT-0500 (Central Daylight Time)
```
But, as you pointed out, if I leave the code as it is, it fails for input like `August 19, 1975 23:15:30 GMT+07:00`.  This is because in the openrosa-xpath-evaluator implementation, that input will just fall through to `return new Date(r);`, but in my new code, `rMoment.isValid()` returns true for that input and the time values are stripped off.  Now, this kind of seems like the right behavior since we are acting more consistently when it comes to clearing out the time values.  But this is a behavior change from the old implementation.

I am going to re-add the DATE_STRING test which should make this behave the same as the original implementation in for any input (only stripping the time values from dates that match DATE_STRING), but let me know if you think it would be better to tweak this behavior (so that we always trim the times or never trim them).....